src/rpc/mining.cpp

```diff
@@ -479,9 +479,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             {
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
+                    LEAVE_CRITICAL_SECTION(lock);
```

Comment:
This is a pretty weird construction; generally you shouldn't be releasing the lock that protects the state you're waiting to change while inspecting it.

What about making `CTxMemPool::nTransactionsUpdated` an atomic, and making `CTxMemPool::GetTransactionsUpdated` not grap `cs`? Would it still be needed to release `lock` here then?

Would you be able to write this as `g_best_block_cv.wait_until(lock, checktxtime, [&](){ return g_best_block != hashWatchedChain || !IsRPCRunning() || mempool.GetTransactionsUpdate() != nTransactionsUpdatedLastLP; });`?

Reply:
Made it atomic, but I left the refactoring change for a future pull request.